### DISCUSSION TITLE
Support for legacy Python versions

### DISCUSSION THREAD

**Discussion by kamakazikamikaze** [easysnmp/easysnmp#135] [Announcements]
Everyone,

As I push forward to refresh the codebase, I'd like to open discussion regarding legacy support for now-outdated versions of Python. Specifically, the following cannot be supported via Github Actions with `actions/setup-python@v2`:

* 2.6.x
* 3.3.x
* 3.4.x

These versions are listed as supported in the `setup.py` file but I cannot rely on publicly-maintained Actions to create the build/test environment needed. Figuring out how to support these versions will likely require a significant amount of time that I am certain I cannot afford to invest right now. Python 2.7 *is* possible to support but creating a local environment isn't the easiest. (I'll document my process in case someone needs to create one for themselves.)

However, as a Sysadmin/Network Engineer/Developer/Penetration Tester who has had to work jobs that supported systems with operating systems far beyond their EOL date, I understand that not everyone has the luxury to upgrade and run the latest OS. Therefore, while I will be removing these versions from the Github Actions, I am open to retaining backwards-compatible code so that locally installing from source is possible. If you have an explicit need for a legacy Python version, please let me know here so I can track who I would need to contact when updates are rolled out. I don't wish to leave anyone behind but I also do not want to hold the majority back.

**Comment by kamakazikamikaze** [comment]
As a side note, if I find that there is a significant number of users/supporters who require a Python version that is not supported by `actions/setup-python@v2`, I am open to the idea of creating a separate Action workflow that will explicitly build for said versions.

**Comment by carlkidcrypto** [comment]
My two cents. Drop support for any python versions that are EOL (End of Life). If a user requires support for older python versions they can lock to easysnmp 0.2.5 or earlier. This will make it easier to fix bugs, do CI (continuous integration), etc.

**Reply by kamakazikamikaze** [reply]
That may be what will have to happen, but I hope I can at least merge bug fixes and patch memory leaks for a 0.2.6 version that would be a final supported release for all EOL editions of Python.

**Reply by carlkidcrypto** [reply]
Yeah, not a bad idea to do one last bug fixes/maintenance release for legacy python versions. If you need help testing the 0.2.6 release let me know. I have Ubuntu 20 systems using easysnmp to poll dozens of devices.

**Comment by benafischer94** [comment]
I know users who cling to older versions disagree; however, I don't see any reason to keep supporting 2.x especially. At this point anything older than 3.6 is end of life. Spending the time building, and testing build pipelines, and supporting versions of the language that aren't even support seems a bit of a resource drain.

The state of SNMP support in python...leaves a bit to be desired. I'd love to help and contribute. Honestly though, doing so with requirement to make sure that versions of Python that aren't even supported puts me off a bit.

Just my two cents on the matter.

**Reply by kamakazikamikaze** [reply]
Understandable. Eventually support for 2.x will be dropped. It just boils down to the demand. I'm fine with dropping support sometime this year or with a 0.3 release when all major memory issues are resolved.

**Comment by amotl** [comment]
Dear Kent,

first things first: Thank you very much for conceiving and maintaining this excellent library. I think it is perfectly fine to drop support for Python 2, better sooner than later.

With kind regards,
Andreas.
